### PR TITLE
[TFLite] Add support for the usage of the Softmax reference kernel and register it in BuiltinRefOpResolver

### PR DESCRIPTION
**Pull request by Tessil**
Hello,

The PR adds a support for the usage of the Softmax reference kernel and register it in the `BuiltinRefOpResolver` to allow user to run the reference Softmax kernel. 

The preparations and checks are similar to the ones in [TFLite Micro](https://github.com/tensorflow/tensorflow/blob/b3a0ede93c5252ef8728d283ba35c0d5ce2fb150/tensorflow/lite/micro/kernels/softmax_common.cc#L30).

A second commit fix some memory leaks that where present in the activations tests.

Thibaut

### CODE REVIEW

File: tensorflow/lite/kernels/activations.cc [hunk 1]
```diff
@@ -576,42 +577,72 @@ TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
...
-    TF_LITE_ENSURE(context, input->type == kTfLiteInt8 ||
-                                input->type == kTfLiteUInt8 ||
-                                input->type == kTfLiteInt16);
-  } else {
-    TF_LITE_ENSURE_TYPES_EQ(context, input->type, output->type);
-  }
 
   TF_LITE_ENSURE(context, NumDimensions(input) >= 1);
 
+  if (input->type == kTfLiteUInt8 || input->type == kTfLiteInt8 ||
```
Comment (Tessil) [inline tensorflow/lite/kernels/activations.cc:583]: Note that the checks here are similar to the ones in [TFLite Micro](https://github.com/tensorflow/tensorflow/blob/f66782631d276c390732a513cdef6ad91eb4e779/tensorflow/lite/micro/kernels/softmax_common.cc#L35). For int8 -> int8 and int16 -> int16, the scale and zero-point are restricted properly in [operator_property.cc](https://github.com/tensorflow/tensorflow/blob/f66782631d276c390732a513cdef6ad91eb4e779/tensorflow/lite/tools/optimize/operator_property.cc#L939), I'm not sure for the (u)int8 -> int16 and uint8 -> uint8 cases. We could eventually only do these checks when the reference kernels are called to avoid any potential backward incompatibility.

**Comment by abattery** [comment]
@renjie-liu @jianlijianli @thaink could you review this PR?

File: tensorflow/lite/kernels/activations.cc [hunk 2]
```diff
@@ -576,42 +577,72 @@ TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
...
+      input->type == kTfLiteInt16) {
+    if (input->type == kTfLiteInt16) {
+      TF_LITE_ENSURE_TYPES_EQ(context, output->type, kTfLiteInt16);
+      TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+      TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/activations.cc:588]: why this one is 32768 but ln-595 is 65536?

the softmax should output between [0,1], so all of them should be 65536 right?

File: tensorflow/lite/kernels/activations.cc [hunk 3]
```diff
@@ -576,42 +577,72 @@ TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
...
+      if (output->type == kTfLiteInt16) {
+        TF_LITE_ENSURE(
+            context, input->type == kTfLiteInt8 || input->type == kTfLiteUInt8);
+        TF_LITE_ENSURE_EQ(context, output->params.zero_point, -32768);
+        TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
+                            (0.001f * 1.f / 65536));
+      } else if (output->type == kTfLiteInt8) {
+        TF_LITE_ENSURE_TYPES_EQ(context, input->type, kTfLiteInt8);
+        TF_LITE_ENSURE_EQ(context, output->params.zero_point, -128);
+        TF_LITE_ENSURE(context, output->params.scale == 1.f / 256);
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/activations.cc:600]: this should be ENSURE_NEAR as well

File: tensorflow/lite/kernels/activations.cc (hunk 2 above)
Comment (Tessil) [reply tensorflow/lite/kernels/activations.cc:588]: The zero-point is null for this one contrary to the 65536 one which has a zero-point of -32768.

Comment (renjie-liu) [reply tensorflow/lite/kernels/activations.cc:588]: that confused me even more, I can understand that int16 ranges from -32768 to 32767, and if we minus the zero_point (-32768) and times the output scale, we got a range of [0, 1].

but it seems for this case, we will have a range of [-1, 1]?

File: tensorflow/lite/kernels/activations.cc (hunk 3 above)
Comment (Tessil) [reply tensorflow/lite/kernels/activations.cc]: Fixed it, thanks

File: tensorflow/lite/kernels/activations.cc (hunk 2 above)
Comment (Tessil) [reply tensorflow/lite/kernels/activations.cc:588]: It's mainly because all int16 tensors have a null zero-point due to hardware constraints (constraint is checked also in the other kernels). With a null zero-point we can effectively represent a [-1, 1] range but it isn't a problem in itself as we need to preserve the null zero-point for int16 kernels.

I'm not sure about the (u)int8 -> int16 case and why it's an exception to these constraints, I used the same check as TFLite Micro but someone it would be better double check if it makes sense. How do we even generate models with such operators?

Comment (renjie-liu) [reply tensorflow/lite/kernels/activations.cc:588]: but softmax should have [0, 1] outputs right?

Comment (Tessil) [reply tensorflow/lite/kernels/activations.cc:588]: Yes but even if the int16 Tensor theoretically allows [-1, 1], the output will still be [0, 1], all int16 quantized values in the output tensor will just be >= 0.

**Review by renjie-liu — APPROVED** [review]

**Review by jianlijianli — COMMENTED** [review]
Hi Tessil, thanks for the PR and looks good in general. Please take a look at the comments below.

File: tensorflow/lite/kernels/activations.cc (hunk 1 above)
Comment (Tessil) [reply tensorflow/lite/kernels/activations.cc]: Thank you for the information regarding the (u)int8 -> int16 path. I removed the checks and now only enforce the int8 -> int8 and int16 -> int16 cases.

**Comment by Tessil** [comment]
@gbaned Fixed it, thanks.

**Review by jianlijianli — APPROVED** [review]
Thanks Tessil for updating the PR.

File: tensorflow/lite/kernels/activations.cc [hunk 4]
```diff
@@ -1036,64 +1051,56 @@ TfLiteStatus SigmoidEval(TfLiteContext* context, TfLiteNode* node) {
...
+  }
 #ifdef TFLITE_SOFTMAX_USE_UINT16_LUT
-  optimized_ops::SoftmaxInt8LUT(
-      data->params, GetTensorShape(input), GetTensorData<uint8_t>(input),
-      GetTensorShape(output), GetTensorData<uint8_t>(output));
-#else
-  optimized_ops::Softmax(data->params, GetTensorShape(input),
-                         GetTensorData<uint8_t>(input), GetTensorShape(output),
-                         GetTensorData<uint8_t>(output));
+  else if ((std::is_same<In, uint8_t>::value &&
```
Comment (abattery) [inline tensorflow/lite/kernels/activations.cc:1081]: This is not a good style. Please do not mix the "else" condition and #ifdef macro.

```
#ifdef TFLITE_SOFTMAX_USE_UINT16_LUT
if ((std::is_same<In, uint8_t>::value && ...
```

File: tensorflow/lite/kernels/activations.cc [hunk 5]
```diff
@@ -1036,64 +1051,56 @@ TfLiteStatus SigmoidEval(TfLiteContext* context, TfLiteNode* node) {
...
+            std::is_same<Out, uint8_t>::value) ||
+           (std::is_same<In, int8_t>::value &&
+            std::is_same<Out, int8_t>::value)) {
+    optimized_ops::SoftmaxInt8LUT(
+        data->params, GetTensorShape(input), GetTensorData<In>(input),
+        GetTensorShape(output), GetTensorData<Out>(output));
+  }
 #endif
+  else {
```
Comment (abattery) [inline tensorflow/lite/kernels/activations.cc:1090]: Please do not mix the "else" condition and #ifdef macro.

```
#endif
if (some conditions) {
```

File: tensorflow/lite/kernels/activations.cc (hunk 4 above)
Comment (Tessil) [reply tensorflow/lite/kernels/activations.cc]: I found it more readable like that but I made the change. I restored the functions as they were and just added the reference case to them.

**Comment by gbaned** [comment]
@Tessil  Can you please resolve conflicts? Thanks!

**Comment by Tessil** [comment]
@gbaned Done, thanks.

**Comment by abattery** [comment]
The change is landed at https://github.com/tensorflow/tensorflow/commit/688eba201334706e9bb3e4eed9211dd29d3f64d1.

**Comment by abattery** [comment]
@Tessil if you have any leftover change after the above change, how about uploading a separate PR and closing this PR?

**Comment by Tessil** [comment]
@abattery the merged changes look good. I will close this PR then, thanks!
